### ISSUE TITLE
Fuse the application of attention mask

### ISSUE THREAD

**Issue by Birch-san** [huggingface/diffusers#1891]
When attention mask is applied: it does so by addition. But there's actually an addition we can get for free during the `baddbmm` above that we're not currently using.

I tried fusing the attention masking into the `baddbmm`:  
https://github.com/Birch-san/diffusers/commit/f55dbf312f22fc514f7983599be88319d9b3e928

I've been unable to verify whether this is equivalent, because I'm not actually convinced attention masking _works_:  
https://github.com/huggingface/diffusers/issues/1890

**Comment by Birch-san** [comment]
I'd have to time it to check (there's no PyTorch profiler for Mac). if it's anything like when I fused the _multiply_ (by `self.scale`) into the `baddbmm`: it _could_ be a dramatic difference. but maybe fusing adds will be less dramatic.

is there any use-case where masks are used in self-attention? that's where the difference would be more dramatic (cross-attention is a far smaller task).

**Comment by Birch-san** [comment]
UnCLIP makes use of attention masking, so could be a good subject for comparing perf, but I can't run it on Mac:  
https://github.com/huggingface/diffusers/issues/1791

**Comment by williamberman** [comment]
Nice @Birch-san! Makes a ton of sense. Could you put up your changes in a PR? It'd be nice to merge after I get the additional unCLIP tests added.

**Comment by patil-suraj** [comment]
Bumping this issue. It would be cool to add this.

**Comment by williamberman** [comment]
put up here, thanks @Birch-san!
